server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java

```diff
@@ -39,71 +44,162 @@ public class DfsPhase {
+                searcher.createWeight(context.searcher().rewrite(query), ScoreMode.COMPLETE, 1);
             }
+        }
 
-            context.dfsResult()
-                .termsStatistics(terms, termStatistics)
-                .fieldStatistics(fieldStatistics)
-                .maxDoc(context.searcher().getIndexReader().maxDoc());
-
-            // If kNN search is requested, perform kNN query and gather top docs
-            SearchSourceBuilder source = context.request().source();
-            if (source != null && source.knnSearch() != null) {
-                SearchExecutionContext searchExecutionContext = context.getSearchExecutionContext();
-                KnnSearchBuilder knnSearch = source.knnSearch();
-
-                KnnVectorQueryBuilder knnVectorQueryBuilder = knnSearch.toQueryBuilder();
-                if (context.request().getAliasFilter().getQueryBuilder() != null) {
-                    knnVectorQueryBuilder.addFilterQuery(context.request().getAliasFilter().getQueryBuilder());
+        Term[] terms = stats.keySet().toArray(new Term[0]);
+        TermStatistics[] termStatistics = new TermStatistics[terms.length];
+        for (int i = 0; i < terms.length; i++) {
+            termStatistics[i] = stats.get(terms[i]);
+        }
+
+        context.dfsResult()
+            .termsStatistics(terms, termStatistics)
+            .fieldStatistics(fieldStatistics)
+            .maxDoc(context.searcher().getIndexReader().maxDoc());
+    }
+
+    private void executeKnnVectorQuery(SearchContext context) throws IOException {
+        SearchSourceBuilder source = context.request().source();
+        if (source == null || source.knnSearch() == null) {
+            return;
+        }
+
+        if (context.getProfilers() == null) {
+            executeKnnVectorQueryWithoutProfiling(context);
+        } else {
+            executeKnnVectorQueryWithProfiling(context);
+        }
+    }
+
+    private void executeKnnVectorQueryWithoutProfiling(SearchContext context) throws IOException {
+        KnnSearchBuilder knnSearch = context.request().source().knnSearch();
+        IndexSearcher searcher = new IndexSearcher(context.searcher().getIndexReader());
+        Query query = buildKnnVectorQuery(context);
+
+        TopDocs topDocs = searcher.search(query, knnSearch.k());
+
+        DfsKnnResults knnResults = new DfsKnnResults(topDocs.scoreDocs);
+        context.dfsResult().knnResults(knnResults);
+    }
+
+    private void executeKnnVectorQueryWithProfiling(SearchContext context) throws IOException {
+        long knnStartTime = System.nanoTime();
+
+        KnnSearchBuilder knnSearch = context.request().source().knnSearch();
+
+        Timer rewriteTimer = new Timer();
+        Timer createWeightTimer = new Timer();
+
+        IndexSearcher searcher = new IndexSearcher(context.searcher().getIndexReader()) {
+            @Override
+            public Query rewrite(Query original) throws IOException {
+                rewriteTimer.start();
```

Comment:
Maybe we can check for cancellation here like we do when gathering term statistics? It doesn't seem critical, but is nice for consistency.

Reply:
Good catch, I added these.